peregrine/blueprints/datasets.py

```diff
@@ -0,0 +1,63 @@
+import flask
+import os
+import re
+
+import peregrine.utils
+from peregrine.resources.submission import (
+    graphql,
+    set_read_access_projects_for_public_endpoint,
+    set_read_access_projects,
+)
+
+from cdiserrors import UserError, AuthZError
+
+blueprint = flask.Blueprint("datasets", "datasets")
+
+
+@blueprint.route("/", methods=["GET"])
+def get_datasets():
+    """
+    Get dataset level summary counts, if a deployment is configured
+    to set PUBLIC_DATASETS to True, this endpoint will be open to
+    anonymous users
+    """
+    nodes = flask.request.args.get("nodes", "")
+    if not nodes:
+        raise UserError("Need to provide target nodes in query param")
+    nodes = nodes.split(",")
+    if os.environ.get("PUBLIC_DATASETS", "").lower() == "true":
+        set_read_access_projects_for_public_endpoint()
+    else:
+        set_read_access_projects()
+    projects = flask.g.read_access_projects
+    if not projects:
+        raise AuthZError("You are not authorized to access any projects")
+    # construct a query that get counts for all projects
+    # because graphql can't add structure to group by projects,
+    # we labeled the count by project index and later parse it
+    # with regex to add structure to response
+    query = "{"
+    for i, project_id in enumerate(projects):
+        query += (
+            " ".join(
+                map(
+                    lambda x: """i{i}_{node}: _{node}_count(project_id: "{p}")""".format(
+                        i=i, node=x, p=project_id
+                    ),
+                    nodes,
+                )
+            )
+            + " "
+        )
+    query += "}"
+    data, errors = graphql.execute_query(query, variables={})
+    if errors:
+        return flask.jsonify({"data": data, "errors": errors}), 400
+    result = {project_id: {} for project_id in projects}
+
+    for name, value in data.iteritems():
+        match = re.search("^i(\d)_(.*)", name)
+        index = int(match.group(1))
+        node = match.group(2)
+        result[projects[index]][node] = value
+    return flask.jsonify(result)
```

Comment:
would it be safer to add `set_read_access_projects()` at the end?
right now `/graphql` does it anyway and it's the only other non-public endpoint, but what if we add another endpoint later

Reply:
actually `set_read_access_projects()` does not do updates
https://github.com/uc-cdis/peregrine/blob/e189498b7fed9bfdc42a8807a1f1752cacd05897/peregrine/resources/submission/__init__.py#L87
so what happens if you request public datasets and then do a normal query? would you now have access to all projects?

Reply:
flask.g is per request